sound/soc/sof/topology.c

```diff
@@ -2307,49 +2311,23 @@ static int sof_complete(struct snd_soc_component *scomp)
 {
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
 	struct snd_sof_widget *swidget, *comp_swidget;
-	struct snd_sof_route *sroute;
-	int ret;
+	const struct topology_widget_ipc_ops *tplg_widget_ipc_ops;
+	const struct ipc_tplg_ops *ipc_tplg_ops;
+	int ret = 0;
+
+	ipc_tplg_ops = sdev->ipc_ops->tplg_ipc_ops;
+	tplg_widget_ipc_ops = ipc_tplg_ops->tplg_widget_ipc_ops;
 
 	/* now update all ipc based on the widget tuple array for host components */
 	list_for_each_entry(swidget, &sdev->widget_list, list) {
-		switch (swidget->id) {
-		case snd_soc_dapm_aif_in:
-		case snd_soc_dapm_aif_out:
-			ret = sof_widget_update_ipc_comp_host(scomp, swidget);
-			break;
-		case snd_soc_dapm_buffer:
-			ret = sof_widget_update_ipc_comp_buffer(scomp, swidget);
-			break;
-		case snd_soc_dapm_mixer:
-			ret = sof_widget_update_ipc_comp_mixer(scomp, swidget);
-			break;
-		case snd_soc_dapm_siggen:
-			ret = sof_widget_update_ipc_comp_tone(scomp, swidget);
-			break;
-		case snd_soc_dapm_scheduler:
-			ret = sof_widget_update_ipc_comp_pipeline(scomp, swidget);
-			break;
-		case snd_soc_dapm_src:
-			ret = sof_widget_update_ipc_comp_src(scomp, swidget);
-			break;
-		case snd_soc_dapm_asrc:
-			ret = sof_widget_update_ipc_comp_asrc(scomp, swidget);
-			break;
-		case snd_soc_dapm_mux:
-		case snd_soc_dapm_demux:
-			ret = sof_widget_update_ipc_comp_mux(scomp, swidget);
-			break;
-		case snd_soc_dapm_pga:
-			ret = sof_widget_update_ipc_comp_pga(scomp, swidget);
-			break;
-		case snd_soc_dapm_effect:
-			ret = sof_widget_update_ipc_comp_process(scomp, swidget);
-			break;
-		case snd_soc_dapm_dai_in:
-		case snd_soc_dapm_dai_out:
-			ret = sof_widget_update_ipc_comp_dai(scomp, swidget);
-			break;
-		default:
+		int i = 0;
+
+		for (i = 0; i < ipc_tplg_ops->tplg_widget_ops_count; i++) {
+			if (swidget->id != tplg_widget_ipc_ops[i].widget_type)
+				continue;
```

Comment:
should we aim for performance with a small hit on memory use by something along this:

```patch
diff --git a/sound/soc/sof/ipc3-topology.c b/sound/soc/sof/ipc3-topology.c
index 0ac5d4f2b0ce..ccb8ba64b331 100644
--- a/sound/soc/sof/ipc3-topology.c
+++ b/sound/soc/sof/ipc3-topology.c
@@ -1140,26 +1140,25 @@ void sof_ipc3_comp_dai_unload(struct snd_soc_component *scomp, struct snd_sof_wi
 	}
 }
 
-const struct topology_widget_ipc_ops tplg_ipc3_widget_ops[] = {
-	{snd_soc_dapm_aif_in, sof_widget_update_ipc_comp_host, NULL},
-	{snd_soc_dapm_aif_out, sof_widget_update_ipc_comp_host, NULL},
-	{snd_soc_dapm_dai_in, sof_widget_update_ipc_comp_dai, sof_ipc3_comp_dai_unload},
-	{snd_soc_dapm_dai_out, sof_widget_update_ipc_comp_dai, sof_ipc3_comp_dai_unload},
-	{snd_soc_dapm_buffer, sof_widget_update_ipc_comp_buffer, NULL},
-	{snd_soc_dapm_mixer, sof_widget_update_ipc_comp_mixer, NULL},
-	{snd_soc_dapm_src, sof_widget_update_ipc_comp_src, NULL},
-	{snd_soc_dapm_asrc, sof_widget_update_ipc_comp_asrc, NULL},
-	{snd_soc_dapm_siggen, sof_widget_update_ipc_comp_tone, NULL},
-	{snd_soc_dapm_scheduler, sof_widget_update_ipc_comp_pipeline, NULL},
-	{snd_soc_dapm_pga, sof_widget_update_ipc_comp_pga, NULL},
-	{snd_soc_dapm_mux, sof_widget_update_ipc_comp_mux, NULL},
-	{snd_soc_dapm_demux, sof_widget_update_ipc_comp_mux, NULL},
-	{snd_soc_dapm_effect, sof_widget_update_ipc_comp_process, NULL},
+const struct topology_widget_ipc_ops tplg_ipc3_widget_ops[SND_SOC_DAPM_TYPE_COUNT] = {
+	[snd_soc_dapm_aif_in] = {sof_widget_update_ipc_comp_host, NULL},
+	[snd_soc_dapm_aif_out] = {sof_widget_update_ipc_comp_host, NULL},
+	[snd_soc_dapm_dai_in] = {sof_widget_update_ipc_comp_dai, sof_ipc3_comp_dai_unload},
+	[snd_soc_dapm_dai_out] = {sof_widget_update_ipc_comp_dai, sof_ipc3_comp_dai_unload},
+	[snd_soc_dapm_buffer] = {sof_widget_update_ipc_comp_buffer, NULL},
+	[snd_soc_dapm_mixer] = {sof_widget_update_ipc_comp_mixer, NULL},
+	[snd_soc_dapm_src] = {sof_widget_update_ipc_comp_src, NULL},
+	[snd_soc_dapm_asrc] = {sof_widget_update_ipc_comp_asrc, NULL},
+	[snd_soc_dapm_siggen] = {sof_widget_update_ipc_comp_tone, NULL},
+	[snd_soc_dapm_scheduler] = {sof_widget_update_ipc_comp_pipeline, NULL},
+	[snd_soc_dapm_pga] = {sof_widget_update_ipc_comp_pga, NULL},
+	[snd_soc_dapm_mux] = {sof_widget_update_ipc_comp_mux, NULL},
+	[snd_soc_dapm_demux] = {sof_widget_update_ipc_comp_mux, NULL},
+	[snd_soc_dapm_effect] = {sof_widget_update_ipc_comp_process, NULL},
 };
 
 const struct ipc_tplg_ops ipc3_tplg_ops = {
 	.tplg_widget_ipc_ops = tplg_ipc3_widget_ops,
-	.tplg_widget_ops_count = ARRAY_SIZE(tplg_ipc3_widget_ops),
 	.route_ipc = NULL,
 };
 
diff --git a/sound/soc/sof/sof-audio.h b/sound/soc/sof/sof-audio.h
index e797bb0cb40d..b8cf047768e0 100644
--- a/sound/soc/sof/sof-audio.h
+++ b/sound/soc/sof/sof-audio.h
@@ -35,14 +35,12 @@ struct snd_sof_widget;
 struct snd_sof_route;
 
 struct topology_widget_ipc_ops {
-	int widget_type;
 	int (*ipc_update)(struct snd_soc_component *scomp, struct snd_sof_widget *swidget);
 	void (*widget_unload)(struct snd_soc_component *scomp, struct snd_sof_widget *swidget);
 };
 
 struct ipc_tplg_ops {
 	const struct topology_widget_ipc_ops *tplg_widget_ipc_ops;
-	int tplg_widget_ops_count;
 	int (*route_ipc)(struct snd_soc_component *scomp, struct snd_sof_route *sroute);
 };
 
diff --git a/sound/soc/sof/topology.c b/sound/soc/sof/topology.c
index 036b55b23093..aabafe3f0b71 100644
--- a/sound/soc/sof/topology.c
+++ b/sound/soc/sof/topology.c
@@ -1870,15 +1870,9 @@ static int sof_widget_unload(struct snd_soc_component *scomp,
 		dai = swidget->private;
 
 		if (dai) {
-			int i = 0;
+			if (tplg_widget_ipc_ops[swidget->id].widget_unload)
+				tplg_widget_ipc_ops[swidget->id].widget_unload(scomp, swidget);
 
-			for (i = 0; i < ipc_tplg_ops->tplg_widget_ops_count; i++) {
-				if (swidget->id != tplg_widget_ipc_ops[i].widget_type)
-					continue;
-
-				tplg_widget_ipc_ops[i].widget_unload(scomp, swidget);
-				break;
-			}
 			list_del(&dai->list);
 		}
 		break;
@@ -2321,16 +2315,9 @@ static int sof_complete(struct snd_soc_component *scomp)
 
 	/* now update all ipc based on the widget tuple array for host components */
 	list_for_each_entry(swidget, &sdev->widget_list, list) {
-		int i = 0;
-
-		for (i = 0; i < ipc_tplg_ops->tplg_widget_ops_count; i++) {
-			if (swidget->id != tplg_widget_ipc_ops[i].widget_type)
-				continue;
+		if (tplg_widget_ipc_ops[swidget->id].ipc_update)
+			ret = tplg_widget_ipc_ops[swidget->id].ipc_update(scomp, swidget);
 
-			if (tplg_widget_ipc_ops[i].ipc_update)
-				ret = tplg_widget_ipc_ops[i].ipc_update(scomp, swidget);
-			break;
-		}
 		if (ret < 0) {
 			dev_err(sdev->dev, "failed updating IPC struct for %s\n",
 				swidget->widget->name);


```

Reply:
@ujfalusi this is a nice optimization but I want to reserve this for the future if you dont mind. At the moment, this part is done only once during topology parsing and so we have a bit of luxury wrt performance.